File: Training/code/Utils/rigid_body.py
import numpy as np
from pyquaternion import Quaternion
from Utils import euler_angle
import math

def quaternion_to_orientation_vector(quat):
    return quat.rotation_matrix.dot(np.array([0, 0, -1]))

def wrap2PI(radians):
    for i in range(len(radians)):
        while radians[i] < -math.pi:
            radians[i] += 2.0 * math.pi
        while radians[i] > math.pi:
            radians[i] -= 2.0 * math.pi
    return radians

class RigidBody:
    def __init__(self, mass, inertia_body, 
        position = np.zeros(3), velocity = np.zeros(3), 
        orientation = Quaternion(axis = [0, 0, 1], angle = 0), 
        angular_velocity = np.zeros(3)):

        self.dt = 0.05
        self.mass = mass
        self.inertia_body = inertia_body
        self.inertia_body_inv = np.linalg.inv(inertia_body)
        self.position = position
        self.velocity = velocity
        self.orientation = orientation

        self.linear_momentum = velocity * mass
        self.R = orientation.rotation_matrix # R is the rotation matrix from body frame to world frame
        self.omega = angular_velocity
        self.angular_momentum = self.R.dot(self.inertia_body).dot(self.R.transpose()).dot(self.omega)
        self.inertia_inv = self.R.dot(self.inertia_body_inv).dot(self.R.transpose())
        self.acceleration = np.zeros(3)
        self.external_force = np.zeros(3)
        self.external_torque = np.zeros(3)

        self.omega_body = self.R.transpose().dot(self.omega)
        self.velocity_body = self.R.transpose().dot(self.velocity)
        self.angle_axis = self.orientation.angle * self.orientation.axis
        self.rpy = euler_angle.quaternion_to_rpy(self.orientation)
        self.rpy = wrap2PI(self.rpy)
        self.rpy_rate = euler_angle.compute_rpy_rate(self.rpy, np.linalg.inv(self.R).dot(self.omega))

        self.orientation_vector = quaternion_to_orientation_vector(self.orientation)

    def wrap2PI(self, x):
        while x > math.pi:
            x -= math.pi * 2.0
        while x < -math.pi:
            x += math.pi * 2.0
        return x

    def apply_force_in_world_frame(self, force, position):
        self.external_force += force
        self.external_torque += np.cross(position - self.position, force)
    
    def apply_force_in_body_frame(self, force, position):
        self.apply_force_in_world_frame(self.vector_body_to_world(force), self.point_body_to_world(position))

    def apply_torque(self, torque):
        self.external_torque += torque
    
    def clear(self):
        self.external_force = np.zeros(3)
        self.external_torque = np.zeros(3)
    
    def advance_one_step(self, dt):
        self.position = self.position + self.velocity * dt
        
        omega = Quaternion(0.0, self.omega[0] * dt, self.omega[1] * dt, self.omega[2] * dt)
        orientation_dot = (omega * self.orientation) * 0.5
        self.orientation = self.orientation + orientation_dot
        self.orientation._normalise()

        self.linear_momentum += self.external_force * dt
        
        self.angular_momentum += self.external_torque * dt

        self.R = self.orientation.rotation_matrix
        self.inertia_inv = self.R.dot(self.inertia_body_inv).dot(self.R.transpose())
        self.velocity = self.linear_momentum / self.mass
        self.acceleration = self.external_force / self.mass
        self.omega = self.inertia_inv.dot(self.angular_momentum)

        self.omega_body = self.R.transpose().dot(self.omega)
        self.velocity_body = self.R.transpose().dot(self.velocity)
        self.angle_axis = self.orientation.angle * self.orientation.axis
        self.rpy = euler_angle.quaternion_to_rpy(self.orientation)
        self.rpy = wrap2PI(self.rpy)
        self.rpy_rate = euler_angle.compute_rpy_rate(self.rpy, np.linalg.inv(self.R).dot(self.omega))
        
        self.orientation_vector = quaternion_to_orientation_vector(self.orientation)

    def advance(self, dt):
        T = 0.0
        while T < dt:
            if self.dt < dt - T:
                one_step_dt = self.dt
            else:
                one_step_dt = dt - T
            self.advance_one_step(one_step_dt)
            T += one_step_dt
        self.clear()

    def vector_body_to_world(self, body_vector):
        return self.R.dot(body_vector)
    
    def vector_world_to_body(self, world_vector):
        return self.R.transpose().dot(world_vector)
        
    def point_body_to_world(self, body_point):
        return self.R.dot(body_point) + self.position

File: SimulationUI/projects/copter_simulation/RigidBody/RigidBody.cpp
#include "RigidBody.h"
#include <iostream>
namespace copter_simulation {

    RigidBody::RigidBody() 
    : _mass(1.0),
    _inertia_body(Eigen::Matrix3d::Identity()), 
    _inertia_body_inv(Eigen::Matrix3d::Identity()),
    _position(Eigen::Vector3d::Zero()),
    _orientation(Eigen::Quaterniond::Identity()), 
    _linear_momentum(Eigen::Vector3d::Zero()),
    _angular_momentum(Eigen::Vector3d::Zero()),
    _inertia_inv(Eigen::Matrix3d::Identity()),
    _R(Eigen::Matrix3d::Identity()),
    _velocity(Eigen::Vector3d::Zero()),
    _acceleration(Eigen::Vector3d::Zero()),
    _omega(Eigen::Vector3d::Zero()),
    _external_force(Eigen::Vector3d::Zero()),
    _external_torque(Eigen::Vector3d::Zero()) {}
    
    void RigidBody::Initialize(const double mass, const Eigen::Matrix3d& inertia_body,
        const Eigen::Vector3d& position, const Eigen::Vector3d& velocity,
        const Eigen::Quaterniond& orientation,
        const Eigen::Vector3d& angular_velocity) {
        _mass = mass;
        _inertia_body = Eigen::Matrix3d(inertia_body);
        _inertia_body_inv = _inertia_body.inverse();
        _position = Eigen::Vector3d(position);
        _orientation = Eigen::Quaterniond(orientation);
        _linear_momentum = velocity * mass;
        _R = _orientation.matrix();
        _omega = Eigen::Vector3d(angular_velocity);
        _angular_momentum = _R * _inertia_body * _R.transpose() * _omega;
        _inertia_inv = _R * _inertia_body_inv * _R.transpose();
        _velocity = Eigen::Vector3d(velocity);
        _acceleration.setZero();
        _external_force.setZero();
        _external_torque.setZero();
    }

    void RigidBody::Initialize(const Eigen::Vector3d& position,
        const Eigen::Vector3d& velocity, const Eigen::Quaterniond& orientation,
        const Eigen::Vector3d& angular_velocity) {
        _position = Eigen::Vector3d(position);
        _orientation = Eigen::Quaterniond(orientation);
        _linear_momentum = velocity * _mass;
        _R = _orientation.matrix();
        _omega = Eigen::Vector3d(angular_velocity);
        _angular_momentum = _R * _inertia_body * _R.transpose() * _omega;
        _inertia_inv = _R * _inertia_body_inv * _R.transpose();
        _velocity = Eigen::Vector3d(velocity);
        _acceleration.setZero();
        _external_force.setZero();
        _external_torque.setZero();
    }

    void RigidBody::ApplyForceInWorldFrame(const Eigen::Vector3d& force, const Eigen::Vector3d& position) {
        _external_force += force;
        _external_torque += (position - _position).cross(force);
    }

    void RigidBody::ApplyForceInBodyFrame(const Eigen::Vector3d& force, const Eigen::Vector3d& position) {
        ApplyForceInWorldFrame(VectorBodyToWorld(force), PointBodyToWorld(position));
    }

    void RigidBody::ApplyTorque(const Eigen::Vector3d& torque) {
        _external_torque += torque;
    }

    void RigidBody::Clear() {
        _external_force.setZero();
        _external_torque.setZero();
    }

    void RigidBody::Advance(const double dt) {
        // update states variables first
        _position = _position + _velocity * dt;
        
        Eigen::Quaterniond omega(0.0, _omega(0), _omega(1), _omega(2));
        _orientation = Add(_orientation, Multiply(omega * _orientation, 0.5 * dt));
        _orientation.normalize();

        _linear_momentum += _external_force * dt;

        _angular_momentum += _external_torque * dt;
      
        // update derived quantities
        _R = _orientation.matrix();
        _inertia_inv = _R * _inertia_body_inv * _R.transpose();
        _velocity = _linear_momentum / _mass;
        _acceleration = _external_force / _mass;
        _omega = _inertia_inv * _angular_momentum;

        // clear external input
        Clear();
    }

    const Eigen::Vector3d RigidBody::PointWorldToBody(
        const Eigen::Vector3d& world_point) const {
        return _R.transpose() * (world_point - _position);
    }

    const Eigen::Matrix3Xd RigidBody::PointsWorldToBody(
        const Eigen::Matrix3Xd& world_points) const {
        return _R.transpose() * (world_points.colwise() - _position);
    }

    const Eigen::Vector3d RigidBody::PointBodyToWorld(
        const Eigen::Vector3d& body_point) const {
        return _R * body_point + _position;
    }

    const Eigen::Matrix3Xd RigidBody::PointsBodyToWorld(
        const Eigen::Matrix3Xd& body_points) const {
        return (_R * body_points).colwise() + _position;
    }

    const Eigen::Vector3d RigidBody::VectorWorldToBody(
        const Eigen::Vector3d& world_vector) const {
        return _R.transpose() * world_vector;
    }

    const Eigen::Matrix3Xd RigidBody::VectorsWorldToBody(
        const Eigen::Matrix3Xd& world_vectors) const {
        return _R.transpose() * world_vectors;
    }

    const Eigen::Vector3d RigidBody::VectorBodyToWorld(
        const Eigen::Vector3d& body_vector) const {
        return _R * body_vector;
    }

    const Eigen::Matrix3Xd RigidBody::VectorsBodyToWolrd(
        const Eigen::Matrix3Xd& body_vectors) const {
        return _R * body_vectors;
    }

    const Eigen::Matrix4d RigidBody::BodyToWorldTransform() const {
        Eigen::Matrix4d transform;
        transform.block(0, 0, 3, 3) = _R;
        transform.block(0, 3, 3, 1) = _position;
        transform.row(3) << 0, 0, 0, 1.0;
        return transform;
    }

    const Eigen::Matrix4d RigidBody::WorldToBodyTransform() const {
        Eigen::Matrix4d transform;
        transform.block(0, 0, 3, 3) = _R.transpose();
        transform.block(0, 3, 3, 1) = -_R.transpose() * _position;
        transform.row(3) << 0, 0, 0, 1.0;
        return transform;
    }

    Eigen::Quaterniond RigidBody::Add(const Eigen::Quaterniond& q1,
        const Eigen::Quaterniond& q2) {
        return Eigen::Quaterniond(q1.w() + q2.w(), q1.x() + q2.x(),
            q1.y() + q2.y(), q1.z() + q2.z());
    }

    Eigen::Quaterniond RigidBody::Multiply(const Eigen::Quaterniond& q,
        const double scale) {
        return Eigen::Quaterniond(q.w() * scale, q.x() * scale,
            q.y() * scale, q.z() * scale);
    }
}